### ISSUE TITLE
Restructure the optimization driver interface for v2.0

### ISSUE THREAD

**Issue by gcarleo** [netket/netket#107]
Version 2.0 calls for a re-thinking of the way we handle the output. At the moment, this is done in the (excellent) json output writer developed by @femtobit. 

I think that in 2.0 we should go beyond the json output. 

As a first goal, I propose to go for something similar to what is done in [Keras](https://keras.io/visualization/) , where the training history is saved in python dictionaries. 

```python
import matplotlib.pyplot as plt
... construct various objects here ... 
history = vmc.run(n_iter=300,n_samples=1000,save_every=10)

# Plot energies 
plt.plot(history.history['energy'].mean) #or something similar for plot with error bars 
plt.show()
```

As a second (slightly more ambitious) goal, it would be really cool to have [callbacks](https://keras.io/callbacks/) à la Keras as well. One can use these callback to print each step on a file, or, even better, to visualise the results dynamically. Once the callbacks are in place, live streaming of the results in a browser can be easily achieved using [Bokeh](https://bokeh.pydata.org/en/latest/) (see [here](https://stackoverflow.com/questions/37724660/streaming-two-line-graphs-using-bokeh) for an example). This would be really cool to have. 

@femtobit what do you think about goal 1 for example? Does it look like an easy modification of your output handlers?

**Comment by femtobit** [comment]
I do agree with the goal of providing different forms of output.
When seeing a line like `history = vmc.run(n_iter=300,n_samples=1000,save_every=10)` it makes me wonder whether it makes sense to *return* it from the run function. It would be more useful, I think to be able to monitor the output in real time (like you also suggest with the callbacks).
But that is step 2, I think.

If we just want to load the history after the simulation is finished (i.e., step 1), it is very easy anyway to access the output so I would advise against returning a history object directly from the `driver.run` functions. We could write a convenience function, e.g.,
```python
op = nk.JsonOutputWriter(logfile, statefile)
vmc.run(..., output=op)
history = op.load_all_history()
```
where `load_all_history` would do something like (of course, in reality it'd have to be implemented as part of the pybind11 bindings, but I will write python code here for readability)
```python
import json
class JsonOutputWriter:
    # [...]
    def load_all_history(self):
        return json.load(self.logfilename)
```
`json.load` returns nested dictionaries for JSON objects, so this is already almost the format you suggest. Converting this, e.g., to a [`pandas`](https://pandas.pydata.org/) dataframe or something similar should be doable in a few lines of python as well. This would also achieve your goal 1 with just slight modifications to the existing code.

When we add a callback interface in step 2, we can make `JsonOutputWriter` a part of it or maybe split it into two classes, `JsonLogOutput` and `JsonStateOutput` that can both be added as callbacks, e.g.,
```python
log = nk.callback.JsonLogOutput('file.json')
statelog = nk.callback.JsonStateOutput('state.json', save_every=50)
vmc.run(..., callback=(log, statelog))

history = log.load_all_history()
...
```

What do you think?

**Comment by gcarleo** [comment]
Hi @femtobit, 
thanks, indeed the two goals are somehow interconnected, and in a sense goal 1 is a special case of goal 2. I agree that something like 

```python
op = nk.JsonOutputWriter(logfile, statefile)
vmc.run(..., output=op)
history = op.load_all_history()
```

makes a lot of sense (names to be discussed in another thread :) ) . 
In some sense, there is not even need to separate goal 2, in the sense that the output argument can be a generic Output object with some basic methods defined (say something like push, so that the vmc can use it to push the results).   Then this Output object is basically our callback of point 2... 

I think that in Keras they would do 
```python
history=vmc.run(...)
```

as a convenient default short-cut when you want the output just to be pushing in a list of dictionaries (in this case history).

**Comment by jamesETsmith** [comment]
I think @femtobit's suggestion to try pandas is a good one. It makes saving and loading the data very simple

**Comment by gcarleo** [comment]
Sure, depending on the technical requirements it might be possible to add it as an Output handler... we have to see. I would nonetheless tend to offer different choices, ranging from very simple-minded lists of dicts to more complicated ones... !

**Comment by twesterhout** [comment]
How about looking at this from the FP/maths perspective? The whole optimisation procedure produces a sequence `{X_i}` for `i` in `[0, #steps-1)`. Let's now drop the upper bound on `i` and say that we have a sequence `{X_i}` for `i` in `N` (Natural numbers). Creators of Python decided to call such sequences _generators_! The only question is a generator of what? I'd say something like this:
```python
# The following is for illustration purposes only and should actually be a C++ object
# and attributes should be immutable
class _State(object):
    def __init__(self, psi:numpy.ndarray, observables:Dict[str, complex]):
        self.psi = psi
        for key, value in observables.items():
            setattr(self, key, value)
# Now if
x = _State(...)
# I can do stuff like this
print(x.energy) # Energy is always sampled
print(x.spin_spin_correlator) # Some observable that I asked netket to sample
```
So that the user code could look like this:
```python
from itertools import islice
for i, state in enumerate(islice(netket.GroundStateOptimizerOrSomething(...), 100)):
    print('{}\t{}'.format(i, state.energy), file=my_file)
```
which saves energies at each of the first 100 iterations to `my_file` in a TSV format. The cool thing is that you can even continue optimising if you're not satisfied. Try running the following in python3:
```python
from itertools import islice
my_iterator = iter(range(10))
for i in islice(my_iterator, 3):
    print(i)
# Prints 0 1 2
for i in islice(my_iterator, 3):
    print(i)
# Bang! Prints 3 4 5
```
So as a user you can do stuff like
```python
optimisation_sequence = iter(netket.GroundStateOptimizerOrSomething(...))
state = None
for i, _state in islice(netket.GroundStateOptimizerOrSomething(...), 100):
    state = _state # This is a trick
    # Some code

if my_custom_condition(state):
    # Nah, let's run another 100 iterations
    for _state in islice(netket.GroundStateOptimizerOrSomething(...), 100):
        state = _state
        # A bit more code
```
And believe me, the continuation feature is just the tip of the iceberg! And there's also that thing called performance which is significantly improved since no serialising / parsing is done and no calls to the OS/filesystem are made.

**Comment by femtobit** [comment]
@twesterhout This would move the control of the simulation run from the C++ side to python, since as far as I understand the driver classes would perform single steps and in between return control to python (where the python code may then decide to continue the iteration, returning control back to the C++ side), right?

The downsides I can see with this approach (compared to periodically calling callback functions that would usually be C++ function objects) are

1) possible degradation of performance (however, this might be insignificant if a single optimization step takes long enough, which might be the case now - I remember that VMC spend most of its time in linear algebra routines);

2) stronger dependence on python to do something useful, as the logic for controlling the simulation has to be supplied by the user's code which iterates over the generator (the upside is of course having this kind of control from the python side). At the moment, I think I would prefer an approach in which the control of the simulation can be done in pure C++ with a similar level of comfort, which an interface in which the driver periodically calls a set of C++ (or possibly python, if the user really wants) callbacks.

**Comment by femtobit** [comment]
> And there's also that thing called performance which is significantly improved since no serialising / parsing is done and no calls to the OS/filesystem are made.

At least in my experience, in the typical use case one would want to write all easily accessible information to disk. You never know what data on your simulation run you might need in the future and re-running costly simulations every time your requirements change is no fun. Thus I'd expect people to almost always want to write at least the information contained in the current log to disk anyway. (And if not, nothing stops them from not supplying the corresponding callback.)

Also, writing a small amount of JSON to disk every step is probably not significantly more expensive than the loop-overhead in python (though I did not test this, so I might be wrong).

**Comment by twesterhout** [comment]
> @twesterhout This would move the control of the simulation run from the C++ side to python, since as far as I understand the driver classes would perform single steps and in between return control to python (where the python code may then decide to continue the iteration, returning control back to the C++ side), right?

This would move the main loop (in CS terms) to Python, yes.


> possible degradation of performance (however, this might be insignificant if a single optimization step takes long enough, which might be the case now - I remember that VMC spent most of its time in linear algebra routines);

Yeah, no. How many iterations do we usually perform, a couple thousands? Here's a test
```python
setup = """
import os

def foo():
    x = os.getcwd()
    return x

def bar():
    for i in range(10000):
        foo()
"""

import timeit
timeit.timeit('bar()', setup=setup, number=1000)
```
A call to `bar` takes less than 10ms on my computer. Importing `numpy` takes more :)

> stronger dependence on python to do something useful, as the logic for controlling the simulation has to be supplied by the user's code which iterates over the generator (the upside is of course having this kind of control from the python side). At the moment, I think I would prefer an approach in which the control of the simulation can be done in pure C++ with a similar level of comfort, which an interface in which the driver periodically calls a set of C++ (or possibly python, if the user really wants) callbacks would provide.

The way to implement what I'm trying to describe here is to use iterators. So in C++ it would look like
```cpp
netket::GroundState generator{...};
for (auto const& state : generator) {
    ...
}
```
Looks simple enough to me...

> At least in my experience, in the typical use case one would want to write all easily accessible information to disk. You never know what data on your simulation run you might need in the future and re-running costly simulations every time your requirements change is no fun. Thus I'd expect people to almost always want to write at least the information contained in the current log to disk anyway. (And if not, nothing stops them from not supplying the corresponding callback.)

But writing something to disk is a matter of a single `print` statement inside the loop, isn't it? The main reason for providing Python wrappers is _expressibility_. And that's why moving the main loop to Python is cool. At virtually no overhead for a beginner, a user that actually does something useful with NetKet is given freedom to implement his custom stopping criterion, collect additional statistics and what not.

> Also, writing a small amount of JSON to disk every step is probably not significantly more expensive than the loop-overhead in python (though I did not test this, so I might be wrong).

How about this: converting a 10^6 element vector of complex numbers (currently) requires a creation of a 10^6 `std::vector`s, i.e. a million calls to `malloc`... Also I can always switch to PyPy which provides very decent performance for simple loops.

**Comment by femtobit** [comment]
> The way to implement what I'm trying to describe here is to use iterators. So in C++ it would look like [...] Looks simple enough to me...

Fair enough.

**Comment by twesterhout** [comment]
@gcarleo that's not fair :) If you'd like to keep things simple, then it's more like
```python
from itertools import islice
X = netket.Something(...)
for state in islice(X.run(), 100):
    print(state.iteration, state.energy, ...)
```
vs.
```python
import sys
X = netket.Something(...)
history = {}
X.run(..., output=history)
netket.magically_convert_to_tsv(history, file=sys.stdout)
```
To write iteration numbers and energies to standard output in TSV format.

I think the differences are cosmetic and, if you really want to, there's nothing stopping us from implementing a few python one-liners in the package.

**Comment by gcarleo** [comment]
The other problem is that I don't understand if `energy` etc must be known in advance, i.e. at construction of the object `State` instead at construction of the `vmc` object. I want to be able to measure arbitrary observable that only `vmc` knows, without ever thinking about this `state` object. 
Can't the `state` object be just a dict of lists? 

I agree with @femtobit that in a typical production scenario you never restart the simulation within python, but you save intermediate steps to a file...

**Comment by twesterhout** [comment]
> The other problem is that I don't understand if energy etc must be known in advance, i.e. at construction of the object State instead at construction of the vmc object. I want to be able to measure arbitrary observable that only vmc knows, without ever thinking about this state object.
Can't the state object be just a dict of lists?

@gcarleo I'm afraid I don't understand what you mean... Could explain it in very simple terms for me, please?
The way I see it, instances of the `_State` class can be treated as dictionaries for most practical purposes. But definitely not dictionaries of lists. `state.psi` gives you a numpy array, `state.energy` gives you a complex number, `state.my_custom_observable` gives you a complex number etc. This relies on Python's dynamic attributes.

**Comment by gcarleo** [comment]
That's the thing, I don't understand how you append these attributes to the _State class, and if that's left to the user or not. 
I would like:
```python
vmc=nk.Vmc(...)
vmc.AddObservable(X,"sigmax")
output={}
vmc.Run(output=out)
print(out["sigmax"])  
```

or 
```python
vmc=nk.Vmc(...)
vmc.AddObservable(X,"sigmax")
output=nk.FileOutputWhatever(...)
vmc.Run(output=out)
#output file contains whatever vmc has pushed at each iteration 
```

**Comment by gcarleo** [comment]
Or 
```python
vmc=nk.Vmc(...)
vmc.AddObservable(X,"sigmax")
out=nk.BokehStream("sigmax")
vmc.Run(output=out)
#live stream the plot of the asked observable in a browser 
```

That's the simplicity I am looking for, here at each step the state is actually updated, it's just that you don't return the control to python... Maybe we can have the option of returning the control to python but I think it's too involved as default usage

**Comment by twesterhout** [comment]
> That's the thing, I don't understand how you append these attributes to the _State class, and if that's left to the user or not.

Don't worry, everything is done automatically. `VariationalMonteCarlo` knows which observables to sample, so it can construct the `state` accordingly.

> I would like: [...]

Sure thing, you'd do
```python
vmc = nk.Vmc(...)
vmc.AddObservable(X, "sigmax")
for state in vmc.run():
    # You access sigmax at each iteration using the following syntax
    dummy = state.sigmax
    # or if you prefer spelling it out
    dummy = state['sigmax'] 
```

**Comment by gcarleo** [comment]
I think that the two things are compatible. It's just that your `run` method is actually an `iterate` method, just performing a single iteration. The two methods are different and the user can choose which one he/she prefers

**Comment by twesterhout** [comment]
> That's the simplicity I am looking for, here at each step the state is actually updated, it's just that you don't return the control to python... Maybe we can have the option of returning the control to python but I think it's too involved as default usage

I'm still failing to see what's involved here from the user perspective. We can even have your version of `run` implemented completely in Python without the need to access the browser from C++ (which, if you've ever done it, you know is not so simple ;)).

> It's just that your run method is actually an iterate method, just performing a single iteration. 

Not really. `__next__` performs 1 iteration and that's what the `for` loop is calling under the hood. The `run` method is closer to `return {this->begin, this->end};` in C++.

> The two methods are different and the user can choose which one he/she prefers

As I said, the solution you seem to like is less general and can simply be implemented if the C++ core is designed to support iterators. I understand that implementing custom iterators may seem like advanced stuff, but it's really quite simple in C++11 and above.

**Comment by gcarleo** [comment]
@femtobit what do you think?

**Comment by femtobit** [comment]
I am a bit torn, to be honest. The iterator interface is more general and allows its users more control about the simulation process which might be helpful for advanced use cases, for example with regard to pausing and continuing the simulation.

Note that the callback interface can be implemented in terms of the iterator interface without much effort. Essentially it I think it would boil down to something like this (assuming `vmc.Iter()` provides an iterator over the VMC steps and where `Run(...)` can be passed a callback as in the example above):
```cpp
void Run(..., Callback callback)
  for (const auto& vmc_state : vmc.Iter()) {
    callback(vmc_state);
  }
}
```
Thus, whether one likes the style used by @twesterhout in the examples here or the one @gcarleo prefers, both can be implemented in terms of the iterator interface. The converse is more complicated. A lot of control about the simulation can be provided via callbacks as well (e.g., it is possible to let a callback function signal to the driver whether to continue the simulation run or not) but this needs to be explicitly implemented in the driver and might also require some more work for users to utilize.

However, I do believe this added flexibility comes at the cost of increased complexity for the implementation of the iterators itself while the callback interface is a pretty straightforward generalization of the code and structure we already have in place. Also, I do believe that implementing the driver interface in terms of iterators increases the barrier of entry for future collaborators since it does require a more thorough understanding of C++ concepts (and Python generators) to get right than the callback approach (which mostly requires adding a function call to the callback at the right place in the simulation loop).

Not an easy decision, IMHO.

**Comment by gcarleo** [comment]
Just to add a further point of discussion, notice that we can (and maybe should) also directly expose the inner optimization loop. In c++ it is something like 
```c++
for (int i = 0; i < niter_opt_; i++) {
   Sample();
   Gradient();
   UpdateParameters();
   PrintOutput(i);
}
```

Now since `Sample` `Gradient` and `UpdateParameters` are all part of the iteration, we can just group them together in a single `Iterate` method 

```python
vmc=nk.Vmc(...)
vmc.AddObservable(X,"sigmax")

for i in range(n_iter):
    vmc.Iterate()
    obs=vmc.observables()
    # do whatever you want with obs 
    # including passing them to a streaming service 
```

I think this is the most transparent way of leaving the user with the maximum freedom and flexibility. 
Also, this would be a relatively minor change in our code. I propose we should implement this, and then later think of higher level interfaces like the ones we were discussing before. Since v2.0 is already a major change in the code base, I guess we can leave further high level improvements for v2.1 or something?

**Comment by gcarleo** [comment]
From the performance point of view, I really hope that exposing this high level `for` is not a big deal, otherwise python is worse than I thought.

**Comment by femtobit** [comment]
@gcarleo I'd like to try my hand at implementing the new interface for the iteration along the lines of what we have discussed here. Since for our project in December I will probably need to extend the current VMC code a bit anyways, I think this would actually fit our plans quite well.
However, since at the moment I have several other projects with deadlines approaching quickly, I will probably not get to spend a lot of time on this until next week or, worst case, the week after that. Would that work for you?

**Comment by gcarleo** [comment]
OK, no problem @femtobit !
